node/test/Chain/Abstract.hs

```diff
@@ -0,0 +1,113 @@
+{-# LANGUAGE KindSignatures #-}
+
+-- | An abstract blockchain. This sits between the pure UTxO based DSL and a
+--   full-fat implementation of Cardano.
+
+module Chain.Abstract where
+
+import Pos.Core.Chrono
+import qualified UTxO.DSL as DSL
+import Universum
+import qualified Data.Set as Set
+import qualified Data.Map.Strict as Map
+import qualified Data.Foldable (fold, foldMap)
+import Data.Monoid (Sum(..))
+
+-- | In the abstract DSL, we identify transactions with integers.
+newtype Addr = Addr Int
+  deriving (Eq, Ord, Show)
+
+-- | Function with finite support. Note that the paper defines these as taking
+--   values in a Semiring, but there's nothing intrinsic to that constraint and
+--   Monoid is in the standard libraries.
+class Monoid v => FinitelySupportedFunction f k v | f -> k v where
+  fSupport :: f -> Set.Set k
+  -- | Apply the function. This must return `mempty` when the key is not in the support
+  -- of the function.
+  fApply :: f -> k -> v
+  fSum :: f -> v
+  fSum fn = Data.Foldable.foldMap (fApply fn) . fSupport $ fn
+
+-- | Standard implementation of a finitely supported function as a map.
+instance (Ord k, Monoid v) => FinitelySupportedFunction (Map.Map k v) k v where
+  fSupport = Map.keysSet
+  fApply m k = Map.findWithDefault mempty k m
+  fSum = Data.Foldable.fold . Map.elems
+
+data StakeDistribution a =
```

Comment:
Why does a stake distribution have to be a finitely supported function? Is it so because a stake of 0 is not interesting?

Reply:
Mostly this is just translating from the paper. This is basically saying that only a finite number of people have non-zero stake.